### PR TITLE
Use the local copy of boost library only on x86_64

### PR DESCRIPTION
**Pull request by martin-g**
For any other CPU architecture fall back to the system-wide boost (installed by OS packages)

### CODE REVIEW

**Comment by martin-g** [comment]
I didn't re-create `configure` but edited it manually.
 `autoreconf` modifies it a lot! 
Please let me know if I have to use `autoreconf` and I will update the PR!

File: configure [hunk 1]
```diff
@@ -2851,12 +2851,18 @@ fi
...
 
 #!/bin/sh
-if test -d ./src/boost.tar.gz;
+if test "$(uname -m)" == "x86_64";
   then
-      echo 'found boost header sources and tar archive;\n using what is there.'
-  else
-      echo "untarring boost include tree...";
-      tar -zxf ./src/boost.tar.gz -C ./src;
+        CPPFLAGS="-I. ${CPPFLAGS}"
```
Comment (martin-g) [inline configure]: Add the current folder to the includes path only when it has to use the local copy of boost.
This change is optional but IMO it makes it better. Local header files should be included with `"..."` instead of `<...>`.
Please let me know if you have any concerns and I will remove it!

File: configure [hunk 2]
```diff
@@ -2851,12 +2851,18 @@ fi
 
 
 #!/bin/sh
-if test -d ./src/boost.tar.gz;
```
Comment (martin-g) [inline configure:2854]: IMO there was a bug here. The old code was checking for a folder/directory with the name of the **file**.

File: configure [hunk 3]
```diff
@@ -2851,12 +2851,18 @@ fi
...
+if test "$(uname -m)" == "x86_64";
   then
-      echo 'found boost header sources and tar archive;\n using what is there.'
-  else
-      echo "untarring boost include tree...";
-      tar -zxf ./src/boost.tar.gz -C ./src;
+    CPPFLAGS="-I. ${CPPFLAGS}"
```
Comment (ahmohamed) [inline configure:2856]: Is this change needed? I think we should leave it in `Makevars.in`. It shouldn't make a difference because there will be no headers anyway in `-I.`.

**Comment by ahmohamed** [comment]
> I didn't re-create `configure` but edited it manually. `autoreconf` modifies it a lot! Please let me know if I have to use `autoreconf` and I will update the PR!

No need to re-run `autoconf` IMO, but we should mirror the changes in `configure.ac` for the future.

File: configure (hunk 2 above)
Comment (ahmohamed) [reply configure:2854]: I agree.

**Comment by ahmohamed** [comment]
CI looks good. I think we should be ready. Let me know if you want to revert `-I.` from `Makevars.in` and also mirror the changes in `configure.ac`, otherwise, I can do that after the merge.

One related question, do you know if ARM64 is supported in github actions (an example would be handy)? It would be good to add it to the CI tests.

File: configure (hunk 3 above)
Comment (martin-g) [reply configure:2856]: `-I.` is in the context of `src/` folder and there we have [handlesegfault.h](https://github.com/ahmohamed/NetPathMiner/blob/master/src/handlesegfault.h) and few more.
By removing the `-I.` from Makevars.in and moving it here we add `-I.` only when `src/boost` is used.
If `-I.` is always in Makevars.in then one can do `#include <handlesegfault.h>` by mistake.

**Comment by martin-g** [comment]
> CI looks good. I think we should be ready. Let me know if you want to revert `-I.` from `Makevars.in` and also mirror the changes in `configure.ac`, otherwise, I can do that after the merge.

I explained my reasons at https://github.com/ahmohamed/NetPathMiner/pull/16#discussion_r1122995823 but I will revert this change if you prefer so!

 

About:
> One related question, do you know if ARM64 is supported in github actions (an example would be handy)? It would be good to add it to the CI tests.

Github Actions does not provide github-hosted ARM64 runners.
There are two ways:
1) use [self-hosted runner](https://docs.github.com/en/actions/hosting-your-own-runners/about-self-hosted-runners)
~I haven't tried this myself yet but https://buildjet.com/for-github-actions/blog/hosted-arm-runners-on-github-actions provides free ARM64 runners in their Free plan.~
Otherwise I have written an [article](https://martin-grigorov.medium.com/github-actions-arm64-runner-on-oracle-cloud-a77cdf7a325a) how to use Oracle Cloud VM for free.

2) use https://github.com/uraimo/run-on-arch-action.
It uses QEMU to emulate aarch64 and it is slower. 
See https://github.com/lh3/bwa/blob/139f68fc4c3747813783a488aef2adc86626b01b/.github/workflows/ci.yaml#L23-L46 for example. The drawback is that you need to do everything in the `run` phase: https://github.com/lh3/bwa/blob/139f68fc4c3747813783a488aef2adc86626b01b/.github/workflows/ci.yaml#L45-L46

File: configure (hunk 3 above)
Comment (ahmohamed) [reply configure:2856]: Actually, don't we need those headers to be included? `init.h` and `hme3m.h` are referenced in cpp files.
https://github.com/ahmohamed/NetPathMiner/blob/master/src/PathRanker.cpp#L40

**Comment by ahmohamed** [comment]
Many thanks @martin-g for the CI examples. Really helpful!! I'll see if I can add that after we merge this.

File: configure (hunk 3 above)
Comment (martin-g) [reply configure:2856]: Right! They need to be included!
Currently they are included with `#include "relative/path.h"`.  Note the usage of `"`.
If `-I.` is added then one can mistakenly include them with: `#include <hme3m.h>`. Note the usage of `<` and `>`, i.e. it may load `hme3m.h` from some other location on the INCLUDE_PATH.

Comment (martin-g) [reply configure:2856]: I'll revert it!
I see that this change causes confusion. 
It is an improvement for a hypothetical use case anyway.

Comment (ahmohamed) [reply configure:2856]: Sorry I just noticed the `<>`. I'll resolve this and merge. I'll handle `configure.ac` at a later stage once we can confirm it's building smoothly within the BBS.

**Comment by ahmohamed** [comment]
Thanks again @martin-g for all your efforts. I've merged, pumped version and pushed to upstream.

**Comment by martin-g** [comment]
Thank you, @ahmohamed !
